README.md
# FlutterWidgetExtension
## :large_blue_circle: Internationalization

English | [中文](README_zh.md)

## :information_source: Introductions

A Flutter package for chain build widget, decrease widget nested hell. Only pick the frequently used function. Some method may introduce unnecessary container. Please refre the souce code.

[pub地址](https://pub.dev/packages/flutter_widget_extension)

Example

Before:
```dart
  Widget widget() {
    return Padding(
        padding: const EdgeInsets.all(20.0),
        child: Center(
            child: Stack(
          alignment: AlignmentDirectional.topEnd,
          children: [
            GestureDetector(
              onTap: () => debugPrint("Widget2 is clicked"),
              child: Container(
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(20.0),
                  color: Colors.purple,
                ),
                child: Container(
                  margin: const EdgeInsets.all(4.0),
                  child: ClipRRect(
                    borderRadius: BorderRadius.circular(20.0),
                    child: Image.network(
                        "https://c-ssl.duitang.com/uploads/item/201609/20/20160920124307_mcZHQ.thumb.1000_0.jpeg"),
                  ),
                ),
              ),
            ),
            IconButton(
                onPressed: () => debugPrint("Icon is clicked"),
                icon: const Icon(
                  Icons.highlight_remove,
                  color: Colors.pink,
                )),
          ],
        )));
  }
```
After:
```yaml
dependencies:
  flutter_widget_extension: ^0.0.2
```
```dart
import 'package:flutter_widget_extension/flutter_widget_extension.dart';

  Widget widget() {
    return Image.network(
            "https://c-ssl.duitang.com/uploads/item/201609/20/20160920124307_mcZHQ.thumb.1000_0.jpeg")
        .round(20.0)
        .addMargin(const EdgeInsets.all(4.0))
        .addBackgroundColor(Colors.purple)
        .round(20.0)
        .addOnTap(() => debugPrint("Widget2 is clicked"))
        .badge(IconButton(
            onPressed: () => debugPrint("Icon is clicked"),
            icon: const Icon(
              Icons.highlight_remove,
              color: Colors.pink,
            )))
        .center()
        .addPadding(const EdgeInsets.all(20.0));
  }
```
## :stuck_out_tongue_winking_eye: Authors

[mamz](https://github.com/mamz)

## :copyright: License

[License MIT](LICENSE)


CHANGELOG.md
## 0.0.1

*  A Flutter package for chain build widget, decrease widget nested hell.

## 0.0.2

*  Modify some method name to avoid conflict.


lib/widget_extension.dart
library flutter_widget_extension;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

extension WidgetExtension on Widget {
  Widget decoration(Decoration decoration) {
    return Container(
      decoration: decoration,
      child: this,
    );
  }

  Widget padding(EdgeInsetsGeometry padding) {
    return Padding(
      padding: padding,
      child: this,
    );
  }

  Widget margin(EdgeInsetsGeometry margin) {
    return Container(
      margin: margin,
      child: this,
    );
  }

  Widget backgroundColor(Color color) {
    return Container(
      color: color,
      child: this,
    );
  }

  Widget alignment(AlignmentGeometry alignment) {
    return Container(
      alignment: alignment,
      child: this,
    );
  }

  Widget center() {
    return Center(
      child: this,
    );
  }

  Widget onTap(GestureTapCallback callback) {
    return GestureDetector(
      onTap: callback,
      child: this,
    );
  }

  Widget hero(Object tag) {
    return Hero(
      tag: tag,
      child: this,
    );
  }

  Widget round(double radius) {
    return ClipRRect(
      borderRadius: BorderRadius.circular(radius),
      child: this,
    );
  }

  Widget stadium() {
    return round(10000.0);
  }

  Widget oval() {
    return ClipOval(
      child: this,
    );
  }

  Widget expand({int flex = 1}) {
    return Expanded(
      flex: flex,
      child: this,
    );
  }

  Widget opacity(double opacity) {
    return Opacity(
      opacity: opacity,
      child: this,
    );
  }

  Widget verticalScrollable() {
    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: this,
    );
  }

  Widget horizontalScrollable() {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: this,
    );
  }

  Widget material() {
    return Material(
      child: this,
    );
  }

  /// Add a badge widget on target. Default to add it to right top corner
  /// * [hMargin] is the horizontal distance to left or right margin decided by [alignment]
  /// * [vMargin] is the vertical distance to top or bottom margin decided by [alignment]
  Widget badge(Widget child,
      {AlignmentDirectional alignment = AlignmentDirectional.topEnd, double? hMargin, double? vMargin}) {
    double top = 0.0;
    double bottom = 0.0;
    double left = 0.0;
    double right = 0.0;
    if (vMargin != null) {
      if (alignment.y < 0.0) {
        top = vMargin;
      } else if (alignment.y > 0.0) {
        bottom = vMargin;
      }
    }
    if (hMargin != null) {
      if (alignment.start < 0.0) {
        left = hMargin;
      } else if (alignment.start > 0.0) {
        right = hMargin;
      }
    }
    return Stack(alignment: alignment, children: <Widget>[
      this,
      Container(
        margin: EdgeInsets.only(left: left, right: right, top: top, bottom: bottom),
        child: child,
      )
    ]);
  }
}

vSpace(double height) => SizedBox(
      height: height,
    );

hSpace(double width) => SizedBox(
      width: width,
    );


lib/flutter_widget_extension.dart
library flutter_widget_extension;

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';

extension WidgetExtension on Widget {
  Widget addDecoration(Decoration decoration) {
    return Container(
      decoration: decoration,
      child: this,
    );
  }

  Widget addPadding(EdgeInsetsGeometry padding) {
    return Padding(
      padding: padding,
      child: this,
    );
  }

  Widget addMargin(EdgeInsetsGeometry margin) {
    return Container(
      margin: margin,
      child: this,
    );
  }

  Widget addBackgroundColor(Color color) {
    return Container(
      color: color,
      child: this,
    );
  }

  Widget addAlignment(AlignmentGeometry alignment) {
    return Container(
      alignment: alignment,
      child: this,
    );
  }

  Widget center() {
    return Center(
      child: this,
    );
  }

  Widget addOnTap(GestureTapCallback callback) {
    return GestureDetector(
      onTap: callback,
      child: this,
    );
  }

  Widget hero(Object tag) {
    return Hero(
      tag: tag,
      child: this,
    );
  }

  Widget round(double radius) {
    return ClipRRect(
      borderRadius: BorderRadius.circular(radius),
      child: this,
    );
  }

  Widget stadium() {
    return round(10000.0);
  }

  Widget oval() {
    return ClipOval(
      child: this,
    );
  }

  Widget expand({int flex = 1}) {
    return Expanded(
      flex: flex,
      child: this,
    );
  }

  Widget addOpacity(double opacity) {
    return Opacity(
      opacity: opacity,
      child: this,
    );
  }

  Widget verticalScrollable() {
    return SingleChildScrollView(
      scrollDirection: Axis.vertical,
      child: this,
    );
  }

  Widget horizontalScrollable() {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: this,
    );
  }

  Widget material() {
    return Material(
      child: this,
    );
  }

  /// Add a badge widget on target. Default to add it to right top corner
  /// * [hMargin] is the horizontal distance to left or right margin decided by [alignment]
  /// * [vMargin] is the vertical distance to top or bottom margin decided by [alignment]
  Widget badge(Widget child,
      {AlignmentDirectional alignment = AlignmentDirectional.topEnd,
      double? hMargin,
      double? vMargin}) {
    double top = 0.0;
    double bottom = 0.0;
    double left = 0.0;
    double right = 0.0;
    if (vMargin != null) {
      if (alignment.y < 0.0) {
        top = vMargin;
      } else if (alignment.y > 0.0) {
        bottom = vMargin;
      }
    }
    if (hMargin != null) {
      if (alignment.start < 0.0) {
        left = hMargin;
      } else if (alignment.start > 0.0) {
        right = hMargin;
      }
    }
    return Stack(alignment: alignment, children: <Widget>[
      this,
      Container(
        margin:
            EdgeInsets.only(left: left, right: right, top: top, bottom: bottom),
        child: child,
      )
    ]);
  }
}

vSpace(double height) => SizedBox(
      height: height,
    );

hSpace(double width) => SizedBox(
      width: width,
    );


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_widget_extension/flutter_widget_extension.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: "Widget Extension Demo",
        home: Scaffold(
          appBar: AppBar(
            title: const Text("Widget Extension Demo"),
          ),
          body: Column(children: [
            Container(
              color: Colors.red,
              width: double.infinity,
              height: 200.0,
              child: widget12(),
            ),
            vSpace(20),
            Container(
              color: Colors.red,
              width: double.infinity,
              height: 200.0,
              child: widget22(),
            ),
          ]).verticalScrollable(),
        ));
  }

  Widget widget11() {
    return Center(
      child: ClipRRect(
        borderRadius: BorderRadius.circular(2000.0),
        child: Container(
          color: Colors.blue,
          padding: const EdgeInsets.all(10.0),
          child: const Text(
            "Hello Flutter!",
            style: TextStyle(
              color: Colors.white,
            ),
          ),
        ),
      ),
    );
  }

  Widget widget12() {
    return const Text("Hello Flutter!",
            style: TextStyle(
              color: Colors.white,
            ))
        .addMargin(const EdgeInsets.all(10.0))
        .addBackgroundColor(Colors.blue)
        .stadium()
        .center();
  }

  Widget widget21() {
    return Padding(
        padding: const EdgeInsets.all(20.0),
        child: Center(
            child: Stack(
          alignment: AlignmentDirectional.topEnd,
          children: [
            GestureDetector(
              onTap: () => debugPrint("Widget2 is clicked"),
              child: Container(
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(20.0),
                  color: Colors.purple,
                ),
                child: Container(
                  margin: const EdgeInsets.all(4.0),
                  child: ClipRRect(
                    borderRadius: BorderRadius.circular(20.0),
                    child: Image.network(
                        "https://c-ssl.duitang.com/uploads/item/201609/20/20160920124307_mcZHQ.thumb.1000_0.jpeg"),
                  ),
                ),
              ),
            ),
            IconButton(
                onPressed: () => debugPrint("Icon is clicked"),
                icon: const Icon(
                  Icons.highlight_remove,
                  color: Colors.pink,
                )),
          ],
        )));
  }

  Widget widget22() {
    return Image.network(
            "https://c-ssl.duitang.com/uploads/item/201609/20/20160920124307_mcZHQ.thumb.1000_0.jpeg")
        .round(20.0)
        .addMargin(const EdgeInsets.all(4.0))
        .addBackgroundColor(Colors.purple)
        .round(20.0)
        .addOnTap(() => debugPrint("Widget2 is clicked"))
        .badge(IconButton(
            onPressed: () => debugPrint("Icon is clicked"),
            icon: const Icon(
              Icons.highlight_remove,
              color: Colors.pink,
            )))
        .center()
        .addPadding(const EdgeInsets.all(20.0));
  }
}
